File: MyProjects/MAP/Assignment 2/src/Repository/Repository.java
package Repository;

import Model.ProgramState;

import java.io.*;
import java.util.ArrayList;
import java.util.List;

public class Repository implements IRepo {
    private List<ProgramState> listPrg = new ArrayList<>();
    private int current = 0;
    @Override
    public void addPrg(ProgramState prg) {
        listPrg.add(prg);
    }
    public ProgramState getCurrentPrg() {
        return listPrg.get(current);
    }

    public void logProgramStateExec(String logFilePath) throws IOException {
        PrintWriter logFile=new PrintWriter(new BufferedWriter(new FileWriter(logFilePath,true)));
        logFile.println(this.getCurrentPrg().toString());
        logFile.println("------------------");
        logFile.close();
    }

    public String toString(){
        StringBuilder sb = new StringBuilder();

        for (ProgramState state : listPrg){
            sb.append(state).append("\n");
        }
        return sb.toString();
    }
}

File: MyProjects/MAP/Assignment 7/Interpreter/Repository/Repository.cs
using System;
using System.Collections.Generic;
using System.Text;
using Interpreter.Model;
using Interpreter.Model.Statements;
using System.IO;

namespace Interpreter.Repository
{
    public class Repository:IRepo
    {
        private List<ProgramState> listPrg = new List<ProgramState>();
        private int current = 0;

        public Repository() { }

        public void addPrg(ProgramState prg)
        {
            listPrg.Add(prg);
        }
        public ProgramState getCurrentPrg()
        {
            return listPrg[current];
        }

        public void logProgramStateExec(string logFilePath)
        {
            //StreamWriter logFile=new StreamWriter(logFilePath,true);
            //logFile.WriteLine(this.getCurrentPrg().ToString());
            //logFile.WriteLine("------------------");
            //logFile.Close();
            ProgramState p = this.getCurrentPrg();
            using (StreamWriter log = File.AppendText("C:\\Users\\pechd\\source\\repos\\Interpreter\\Interpreter\\log.txt"))
            {

                log.WriteLine("Stack:\n");
                foreach (IStmt st in p.Stack)
                {
                    log.WriteLine(st);
                }
                log.WriteLine("SymTable:\n");
                foreach (KeyValuePair<string, int> v in p.SymbolTable)
                {
                    log.WriteLine(v.Key + "->" + v.Value);
                }

                log.WriteLine("Out:");
                foreach (int v in p.Output)
                {
                    log.WriteLine(v);
                }
                log.WriteLine("FileTable:");
                foreach (KeyValuePair<int, StreamReader> kvp in p.FileTab)
                {
                    log.WriteLine(kvp.Key + "->" + kvp.Value);
                }

            }
        }

        public override string ToString()
        {
            StringBuilder sb = new StringBuilder();

            foreach (ProgramState state in listPrg)
            {
                sb.Append(state).Append("\n");
            }
            return sb.ToString();
        }
    }
}
